Final tasks

Голощапова Алла

группа 306

# Задание 1

Задание 1

- Сгенерировать данные, извлеченные из нормального распределения, оценить методом максимального правдоподобия параметры распределения. Сравнить оцененные параметры с исходными.
- Пусть alpha = 0.01. Найти доверительный интервал с уровнем доверия 1 - alpha для параметра sigma^2



In [ ]:
import numpy as np
from scipy.stats import norm
import scipy.stats as stats
from statistics import variance


In [ ]:
sample_1 =  np.random.normal(loc=2, scale=5, size=100) # генерация данных
mean_av, variance_av  = np.mean(sample_1), variance(sample_1)
mean_mle, variance_mle = norm.fit(sample_1) # оценка параметров
print(mean_av, variance_av)
print(f'Метод максимального правдоподобия: 1) оценка среднего: {mean_mle}, 2) оценка дисперсии: {variance_mle}')

alpha = 0.01 # уровень доверия 1 - alpha для параметра sigma^2
inter = stats.t.interval(confidence=1 - alpha, df=len(sample_1)-1, loc=np.mean(sample_1), scale=stats.sem(sample_1))
print('\nДоверительный интервал: ', inter)

2.8130648658431476 22.218181194228205
Метод максимального правдоподобия: 1) оценка среднего: 2.8130648658431476, 2) оценка дисперсии: 4.689989273152544

Доверительный интервал:  (1.575078037264136, 4.051051694422159)


# Задание 2

У кого не сделано -- выполнить все задания из ноутбуков по семинарам 3 и 4. По заданию 3 с семинара 4 сделать выводы о результатах тестов (написать словами)

(3 семинар - ниже, 4 было сделано https://colab.research.google.com/drive/1m-hPs2SYe3tqOg6T79-RDQuUQnCV4egj?usp=sharing )

### Задание 2.1

Проверьте для примера выше случай уровня доверия 0.01 и **0.001**

In [ ]:
import numpy as np
import scipy.stats as stats

data = [77660, 77461, 71211, 80287, 69798, 89570, 66486, 82513, 167335, 63020, 81304, 70325, 70451, 83819, 82742, 97737, 91025, 91025, 78769, 87018, 161702, 59574, 61260, 202998]

# Для уровня доверия 0.01
confidence_001 = stats.t.interval(0.99, df=len(data)-1, loc=np.mean(data), scale=stats.sem(data))
print("Доверительный интервал для уровня доверия 0.01:", confidence_001)

# Для уровня доверия 0.001
confidence_0001 = stats.t.interval(0.999, df=len(data)-1, loc=np.mean(data), scale=stats.sem(data))
print("Доверительный интервал для уровня доверия 0.001:", confidence_0001)


Доверительный интервал для уровня доверия 0.01: (69767.4213664546, 110656.74530021206)
Доверительный интервал для уровня доверия 0.001: (62774.019240785885, 117650.1474258811)


Задание 2.2

Похожим образом можно оценить доверительный интервал для стандартного отклонения. Что для этого нужно сделать? Напишите код.

In [ ]:
alpha = 0.05
n = len(data)
var = np.var(data, ddof = 1)

chi_l, chi_r = stats.chi2.ppf([1 - alpha/2, alpha/2], df = n - 1)

left_bound = np.sqrt((n - 1) * var / chi_l) #  в отличии от дисперсии для стандартного отклонения - корень
right_bound = np.sqrt((n - 1) * var / chi_r)

print(f"Доверительный интервал для стандартного отклонения:  {left_bound, right_bound}")

Доверительный интервал для стандартного отклонения:  (27728.833111002717, 50046.62232186614)


Задание 2.3

Имеется выборка data_1. С помощью критерия Пирсона χ2 проверить, извлечена ли выборка из нормально распределенной генеральной совокупности. На основании полученного p−value сделать соответствующие вывод

In [ ]:
data_1 = np.array([-0.92381148, -7.85423217, -3.84392923,  2.09793605, -6.44121148, -6.90856921,
 -4.86015786, -3.18982485, -2.0663121,  -0.97537333, -7.49210514, 15.10622733,
 -2.24056844, -0.30024337,  3.0771819,   3.87282985,  1.44277686,  3.1350076,
 -0.2087979,  -1.96891233, -8.41806126,  3.48189129,  8.29465749, -0.3019667,
  3.59492783,  1.47310056,  1.45552366,  0.48111867, -6.76769424, -9.21688519,
  5.86148415,  1.49726226, -7.92089753, -1.72609032, -2.56080562,  2.96248152,
 -3.38978766,  5.78684458,  5.63570755, -0.17762366,  2.28412479,  2.08721277,
 -7.8374102,   5.36420193,  0.90097827,  3.77879897, -2.29114273, -4.31106285,
 -3.13829459, -0.81855704])

alpha = 0.05
chis = stats.chisquare(data_1)
print(chis[1])

1.0


Не отвергаем нулевую гипотезу,можно считать, что выборка извлечена из нормально распределенной генеральной совокупности.


# Задание 3



Задание 3

Использовать данные Boston с занятий. Линейную регрессию по этим данным мы уже построили. Теперь требуется проверить условия Гаусса--Маркова. Необходимо:
1. Проверить остатки модели на нормальность, используя
- критерий Колмогорова--Смирнова или, как его еще называют, тест Лиллиефорса (https://www.statsmodels.org/dev/generated/statsmodels.stats.diagnostic.lilliefors.html)
2. Проверить остатки модели на гетероскедастичность, используя
- тест Голдфелда--Квандта (https://www.statsmodels.org/dev/generated/statsmodels.stats.diagnostic.het_goldfeldquandt.html)
- тест Бройша--Пагана (https://www.statsmodels.org/stable/generated/statsmodels.stats.diagnostic.het_breuschpagan.html)
3. Проверить наличие автокорреляции в остатках с помощью
- теста Бройша--Годфри (https://www.statsmodels.org/dev/generated/statsmodels.stats.diagnostic.acorr_breusch_godfrey.html)
- тест Дарбина--Уотсона (https://www.statsmodels.org/dev/generated/statsmodels.stats.stattools.durbin_watson.html) Вернется значение статистики: если оно будет находиться в пределах от 1.5 до 2.5, то автокорреляция в остатках не наблюдается.

Сделать выводы по результатам этих тестов (написать словами) о выполнении условий Гаусса--Маркова. Что можно сказать о качестве модели?

In [ ]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.stats.diagnostic import het_goldfeldquandt, het_breuschpagan, acorr_breusch_godfrey
from statsmodels.stats.diagnostic import normal_ad
from statsmodels.stats.stattools import durbin_watson


In [ ]:
house_data = pd.read_csv('BostonHousing.csv')
X = pd.DataFrame(house_data.iloc[:, 0:-1])
y = house_data['medv']
X = sm.add_constant(X)

model = sm.OLS(y, X).fit()

In [ ]:
# # Проверка остатков модели на нормальность
normal = normal_ad(model.resid)[1]

print(f"Тест Колмогорова-Смирнова: {normal}")

Тест Колмогорова-Смирнова: 2.162183463169403e-25


тк значение вероятности близко к 0, то отклоненяется гипотеза о нормальности

не выполняется условие нормальности Гаусса-Маркова

In [ ]:
# Проверка остатков модели на гетероскедастичность
gf = het_goldfeldquandt(model.resid, X)[1]
bp = het_breuschpagan(model.resid, X)[1]

print(f"Тест Голдфелда-Квандта: {gf}, тест Бройша-Пагана:{bp}")

Тест Голдфелда-Квандта: 7.302071915608971e-12, тест Бройша-Пагана:6.265430719873663e-09


присутствует гетероскедастичность,

не выполняется условие гомоскедастичности Гаусса-Маркова

In [ ]:
# Проверка наличия автокорреляции в остатках
bg_test = acorr_breusch_godfrey(model, nlags=1)
dw_statistic = durbin_watson(model.resid)

print(f"Тест Бройша-Годфри: {bg_test[1]}, nест Дарбина-Уотсона: {dw_statistic}")


Тест Бройша-Годфри: 1.0533841688380185e-25, nест Дарбина-Уотсона: 1.0783751186797201


автокорреляция в остатках может присутствовать и  1.08 не находится между 1.5 и 2.5
след-но, также не выполняется  еще одно условие - отсутсвие автокорреляции - Гаусса-Маркова

Таким образом,  модель не удовлетворяет трем условиям Гаусса–Маркова. Из этого следует вывод, что модель не можем назвать*нормальной линейной* регрессионной моделью.